File: Sample/java/src/main/java/net/nend/sample/java/nativead/NativeV2OnListActivity.java
package net.nend.sample.java.nativead;

import android.app.ListActivity;
import android.content.Context;
import android.graphics.Bitmap;
import android.os.Bundle;
import androidx.annotation.NonNull;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import net.nend.android.NendAdNative;
import net.nend.android.NendAdNativeClient;
import net.nend.sample.java.R;

import java.util.ArrayList;
import java.util.List;

public class NativeV2OnListActivity extends ListActivity {

    private final int NORMAL = 0;
    private final int AD = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        NendAdNativeClient mClient = new NendAdNativeClient(this, 485516, "16cb170982088d81712e63087061378c71e8aa5c");
        ArrayList<Object> list = new ArrayList<>();

        for (int i = 1; i < 50; i++) {
            final NativeFeed feed = new NativeFeed();

            if (i == 10 || i == 20 || i == 30) {
                feed.setType(AD);
                mClient.loadAd(new NendAdNativeClient.Callback() {
                    @Override
                    public void onSuccess(NendAdNative nendAdNative) {
                        feed.setNendAdNative(nendAdNative);
                    }

                    @Override
                    public void onFailure(NendAdNativeClient.NendError nendError) {
                    }
                });
            } else {
                feed.setType(NORMAL);
                feed.setContent("【" + i + "行目】インフィード広告記事部分長いテキスト　インフィード広告記事部分長いテキスト　インフィード広告記事部分長いテキスト");
                feed.setMediaName("メディア名" + i);
                feed.setDate("1970/01/01");
            }
            list.add(feed);
        }

        NativeListAdapter adapter = new NativeListAdapter(getApplicationContext(), 0, list);
        setListAdapter(adapter);

    }

    private class NativeListAdapter extends ArrayAdapter<Object> {
        private LayoutInflater mInflater;
        private List<Object> mObjects;

        NativeListAdapter(Context context, int resource, List<Object> objects) {
            super(context, resource, objects);
            mInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            mObjects = objects;
        }

        @Override
        public int getViewTypeCount() {
            return 2;
        }

        @Override
        public int getItemViewType(int position) {
            return ((NativeFeed) mObjects.get(position)).getType();
        }

        @NonNull
        @Override
        public View getView(final int position, View convertView, @NonNull ViewGroup parent) {
            final ViewHolder holder;

            if (convertView == null) {
                convertView = mInflater.inflate(R.layout.native_v2_list_row, null);
                holder = new ViewHolder();
                holder.title = (TextView) convertView.findViewById(R.id.title);
                holder.contentText = (TextView) convertView.findViewById(R.id.content);
                holder.smallText = (TextView) convertView.findViewById(R.id.small_text);
                holder.thumbnail = (ImageView) convertView.findViewById(R.id.thumbnail);
                convertView.setTag(holder);
            } else {
                holder = (ViewHolder) convertView.getTag();
            }

            switch (getItemViewType(position)) {

                case NORMAL:
                    NativeFeed feed = (NativeFeed) getItem(position);
                    assert feed != null;
                    holder.contentText.setText(feed.getContent());
                    holder.smallText.setText(feed.getDate());
                    holder.title.setText(feed.getMediaName());
                    break;

                case AD:
                    final View adRoot = convertView;
                    final NativeFeed adFeed = (NativeFeed) getItem(position);

                    if (adFeed != null && adFeed.getNendAdNative() != null && adFeed.getAdimage() == null) {
                        adFeed.getNendAdNative().downloadAdImage(new NendAdNative.Callback() {
                            @Override
                            public void onSuccess(final Bitmap bitmap) {
                                adFeed.setAdimage(bitmap);

                                runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        holder.title.setText(adFeed.getNendAdNative().getTitleText());
                                        holder.contentText.setText(adFeed.getNendAdNative().getContentText());
                                        holder.smallText.setText(NendAdNative.AdvertisingExplicitly.AD.getText());
                                        holder.thumbnail.setImageBitmap(adFeed.getAdimage());

                                        adFeed.getNendAdNative().activate(adRoot, holder.smallText);
                                        adFeed.getNendAdNative().setOnClickListener(new NendAdNative.OnClickListener() {
                                            @Override
                                            public void onClick(NendAdNative ad) {
                                                Log.d("onclick", "" + ad);
                                            }
                                        });
                                    }
                                });
                            }

                            @Override
                            public void onFailure(Exception error) {
                            }
                        });
                    }
                    if (adFeed != null && adFeed.getNendAdNative() != null) {
                        holder.title.setText(adFeed.getNendAdNative().getTitleText());
                        holder.contentText.setText(adFeed.getNendAdNative().getContentText());
                        holder.smallText.setText(NendAdNative.AdvertisingExplicitly.AD.getText());
                        holder.thumbnail.setImageBitmap(adFeed.getAdimage());
                    }
                    break;
            }

            return convertView;
        }

        class ViewHolder {
            TextView title;
            TextView contentText;
            TextView smallText;
            ImageView thumbnail;
        }

    }


    private class NativeFeed {

        private int type;
        private String content;
        private String date;
        private String mediaName;
        private Bitmap adimage;
        private NendAdNative nendAdNative;

        int getType() {
            return type;
        }

        void setType(int type) {
            this.type = type;
        }

        public String getContent() {
            return content;
        }

        public void setContent(String content) {
            this.content = content;
        }

        String getDate() {
            return date;
        }

        void setDate(String date) {
            this.date = date;
        }

        String getMediaName() {
            return mediaName;
        }

        void setMediaName(String mediaName) {
            this.mediaName = mediaName;
        }

        Bitmap getAdimage() {
            return adimage;
        }

        void setAdimage(Bitmap adimage) {
            this.adimage = adimage;
        }

        public NendAdNative getNendAdNative() {
            return nendAdNative;
        }

        public void setNendAdNative(NendAdNative nendAdNative) {
            this.nendAdNative = nendAdNative;
        }
    }
}

File: Sample/kotlin/src/main/java/net/nend/sample/kotlin/nativead/NativeV2OnListActivity.kt
package net.nend.sample.kotlin.nativead

import android.content.Context
import android.graphics.Bitmap
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import net.nend.android.NendAdNative
import net.nend.android.NendAdNativeClient
import net.nend.android.NendAdNativeListener
import net.nend.sample.kotlin.R
import net.nend.sample.kotlin.SimpleListActivity
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_API_KEY_SMALL_SQUARE
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_SPOT_ID_SMALL_SQUARE

class NativeV2OnListActivity : SimpleListActivity() {

    private enum class ViewType(val id: Int) {
        NORMAL(0),
        AD(1)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val client = NendAdNativeClient(this,
                NATIVE_SPOT_ID_SMALL_SQUARE, NATIVE_API_KEY_SMALL_SQUARE)
        val list = mutableListOf<Any>()

        for (i in 1..49) {
            val feed = NativeFeed()
            if (i == 10 || i == 20 || i == 30) {
                feed.type = ViewType.AD.id
                client.loadAd(object : NendAdNativeClient.Callback {
                    override fun onSuccess(nendAdNative: NendAdNative) {
                        feed.nendAdNative = nendAdNative
                    }

                    override fun onFailure(nendError: NendAdNativeClient.NendError) {}
                })
            } else {
                feed.type = ViewType.NORMAL.id
                feed.content = "【" + i + "行目】インフィード広告記事部分長いテキスト　インフィード広告記事部分長いテキスト　インフィード広告記事部分長いテキスト"
                feed.mediaName = "メディア名$i"
                feed.date = "1970/01/01"
            }
            list.add(feed)
        }

        val adapter = NativeListAdapter(applicationContext, 0, list)
        instantiateListAdapter(adapter)
    }

    private inner class NativeListAdapter(
            context: Context, resource: Int, private val objects: List<Any>) :
            ArrayAdapter<Any>(context, resource, objects) {

        private val inflater: LayoutInflater =
                context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

        override fun getViewTypeCount() = ViewType.values().size

        override fun getItemViewType(position: Int) = (objects[position] as NativeFeed).type

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val view: View
            val holder: ViewHolder

            if (convertView == null) {
                view = inflater.inflate(R.layout.native_v2_list_row, parent, false)
                holder = ViewHolder().apply {
                    title = view.findViewById(R.id.title) as TextView
                    contentText = view.findViewById(R.id.content) as TextView
                    smallText = view.findViewById(R.id.small_text) as TextView
                    thumbnail = view.findViewById(R.id.thumbnail) as ImageView
                    view.tag = this
                }
            } else {
                holder = convertView.tag as ViewHolder
                view = convertView
            }

            when (getItemViewType(position)) {
                ViewType.AD.id -> {
                    val adFeed = getItem(position) as NativeFeed?
                    adFeed?.let {
                        if (it.nendAdNative != null && it.adImage == null) {
                            it.nendAdNative?.downloadAdImage(object : NendAdNative.Callback {
                                override fun onSuccess(bitmap: Bitmap) {
                                    adFeed.adImage = bitmap
                                    runOnUiThread {
                                        holder.title?.text = it.nendAdNative?.titleText
                                        holder.contentText?.text = it.nendAdNative?.contentText
                                        holder.smallText?.text = NendAdNative.
                                                AdvertisingExplicitly.AD.text
                                        holder.thumbnail?.setImageBitmap(it.adImage)
                                        it.nendAdNative?.activate(view, holder.smallText)
                                        it.nendAdNative?.setNendAdNativeListener(object : NendAdNativeListener {
                                            override fun onImpression(ad: NendAdNative) {
                                                Log.i(TAG, "onImpression")
                                            }

                                            override fun onClickAd(ad: NendAdNative) {
                                                Log.i(TAG, "onClickAd")
                                            }

                                            override fun onClickInformation(ad: NendAdNative) {
                                                Log.i(TAG, "onClickInformation")
                                            }
                                        })
                                    }
                                }

                                override fun onFailure(error: Exception) {}
                            })
                        }
                        holder.smallText?.text = NendAdNative.AdvertisingExplicitly.AD.text
                        holder.thumbnail?.setImageBitmap(it.adImage)
                        adFeed.nendAdNative?.let { nendAd ->
                            holder.title?.text = nendAd.titleText
                            holder.contentText?.text = nendAd.contentText
                        }
                    }
                }
                else -> {
                    val feed = getItem(position) as NativeFeed?
                    feed?.let {
                        holder.contentText?.text = it.content
                        holder.smallText?.text = it.date
                        holder.title?.text = it.mediaName
                    }
                }
            }

            return view
        }

        inner class ViewHolder {
            var title: TextView? = null
            var contentText: TextView? = null
            var smallText: TextView? = null
            var thumbnail: ImageView? = null
        }
    }

    private inner class NativeFeed {
        var type: Int = 0
        var content: String? = null
        var date: String? = null
        var mediaName: String? = null
        var adImage: Bitmap? = null
        var nendAdNative: NendAdNative? = null
    }


    companion object {
        private val TAG = NativeV2OnListActivity::class.java.simpleName
    }

}
